Fix PG.connect keyword arguments deprecation warning on ruby 2.7

nvasilevski opened the pull request:
### Summary

Fixes - https://github.com/rails/rails/issues/44307

As the issue description states, currently, calling `ActiveRecord::Base.postgresql_connection(params)` with a hash prints the `Using the last argument as keyword parameters is deprecated` warning on Ruby 2.7
Or essentially it could be reproduced by calling:
```ruby
ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.new_client({:port=>5432, dbname: "activerecord_unittest"})
``` 

I believe this warning was not addressed mainly because on Ruby 3.0+ `PG.connect`, which is essentially `PG::Connection.new` supports hash as a configuration
https://github.com/ged/ruby-pg/blob/6c436bda9aaf7bc053ef348afd99e2e035d5d0a3/lib/pg.rb#L68-L70
https://github.com/ged/ruby-pg/blob/6c436bda9aaf7bc053ef348afd99e2e035d5d0a3/lib/pg/connection.rb#L651

Also had to tweak the `connection_parameters` initialization in order to support `nil` values, like in this test:

https://github.com/rails/rails/blob/724b733c454a56d8a32be54fdbdf6a0a60379642/activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb#L49-L54

I wonder if it would be worth adding a new test that explicitly tests the `nil` value passed as connection params. Or it's sufficient to have the one mentioned above that tests the behaviour implicitly 🤔 

Feel free to let me know if you think there is a better way to handle `nil` connection params or perhaps we shouldn't support passing `nil` at all? Thanks!

nvasilevski committed:
Fix PG.connect keyword arguments deprecation warning on ruby 2.7

nvasilevski commented:
The failing test seems to be a little broken by itself. I'm going to dig and open a separate PR

matthewd commented:
I think this warning is new since pg 1.3.x was released a week or two ago.

Is this new spelling warning-safe (on our supported Ruby versions) on previous `pg` versions?

nvasilevski commented:
>Is this new spelling warning-safe (on our supported Ruby versions) on previous pg versions?

Oh, this is a good catch, for some reason I completely missed that `1.3.0` is a relatively new version of the gem.
So the signature was changed in `1.3.0`:
https://github.com/ged/ruby-pg/commit/935ecafa5a4f8448291d8b7c3e94b2fe550201f9

But unwrapping a hash still works on older versions, this is what I've checked:

```ruby
# pg_version = "1.2.3"
pg_version = "1.1.4"
gem "pg", "=#{pg_version}"
require "pg"
raise unless PG::VERSION == pg_version
params = {dbname: "activerecord_unittest"}
connection = PG.connect(**params)
connection.exec("select 1;")
```

No warnings related to keyword arguments on older versions of the gem

byroot reviewed (approved):
Yeah, on newer Ruby passing `**hash` works even if the method expect an option hash. In 2.6 and older you'd get:

```ruby
>> def foo(options);end
=> :foo
>> $VERBOSE = true
=> true
>>  foo(**{bar: 1})
(irb):1: warning: in `foo': the last argument was passed as a single Hash
(irb):3: warning: although a splat keyword arguments here
```

But since Rails is now 2.7+ it's all good.

@nvasilevski maybe add a CHANGELOG entry?

nvasilevski commented:
>maybe add a CHANGELOG entry?

Yep, totally, let me add it quickly

Meanwhile would you mind looking on a separate PR that fixes the tests that is currently failing here - https://github.com/rails/rails/pull/44347

nvasilevski commented:
Rebased & added a changelog entry
I'd like to wait for CI to be green, just because previous failure was a bit weird and I wanted to make sure that I've fixed it properly in a separate PR

byroot commented:
The only failures are the ones you fixed in https://github.com/rails/rails/pull/44347, so LGTM.
